src/rgw/driver/sfs/sqlite/dbconn.cc

```diff
@@ -1,81 +1,83 @@
+) {
   try {
-    fs::copy(getDBPath(ctt) + "-wal", get_temporary_db_path(ctt) + "-wal");
-  } catch (const std::filesystem::filesystem_error& e) {
-    if (e.code() != std::errc::no_such_file_or_directory) {
-      throw e;
-    }
+    return storage.pragma.user_version();
+  } catch (const std::system_error& e) {
+    lsubdout(cct, rgw, -1) << "error opening db: " << e.code().message() << " ("
+                           << e.code().value() << "), " << e.what() << dendl;
+    throw e;
   }
+}
 
-  // try to sync the storage based on the temporary db
-  // in case something goes wrong show possible errors and return
-  auto test_storage = _make_storage(get_temporary_db_path(ctt));
-  test_storage.open_forever();
-  test_storage.busy_timeout(5000);
-  bool sync_error = false;
-  std::string result_message;
-  try {
-    auto sync_res = test_storage.sync_schema();
-    std::vector<std::string> non_compatible_tables;
-    for (auto const& [table_name, sync_result] : sync_res) {
-      if (sync_result == orm::sync_schema_result::dropped_and_recreated) {
-        // this result is aggressive as it drops the table and
-        // recreates it.
-        // Data loss is expected and we should warn the user and
-        // stop the final sync in the real database.
-        result_message +=
-            "Table: [" + table_name + "] is no longer compatible. ";
-        non_compatible_tables.push_back(table_name);
-      }
-    }
-    if (non_compatible_tables.size() > 0) {
-      sync_error = true;
-      result_message = "Tables: [ ";
-      for (auto const& table : non_compatible_tables) {
-        result_message += table + " ";
-      }
-      result_message += "] are no longer compatible.";
-    }
-  } catch (std::exception& e) {
-    // check for any other errors (foreign keys constrains, etc...)
-    result_message =
-        "Metadata database might be corrupted or is no longer compatible";
-    sync_error = true;
-  }
-  // remove the temporary db
-  fs::remove(get_temporary_db_path(ctt));
+// example function to upgrade db
+//
+// static void upgrade_metadata_from_v1(sqlite3* db) {
+//   sqlite3_exec(db, "ALTER TABLE foo ...", nullptr, nullptr, nullptr);
+// }
 
-  try {
-    fs::remove(get_temporary_db_path(ctt) + "-wal");
-  } catch (const std::filesystem::filesystem_error& e) {
-    if (e.code() != std::errc::no_such_file_or_directory) {
-      throw e;
+static void upgrade_metadata(
+    CephContext* cct, rgw::sal::sfs::sqlite::Storage& storage, sqlite3* db
+) {
+  while (true) {
+    int cur_version = get_version(cct, storage);
+    ceph_assert(cur_version <= SFS_METADATA_VERSION);
+    ceph_assert(cur_version >= SFS_METADATA_MIN_VERSION);
+    if (cur_version == SFS_METADATA_VERSION) {
+      break;
     }
+
+    // example check to upgrade db
+    // if (cur_version == 1) {
+    //   upgrade_metadata_from_v1(db);
+    // }
+
+    storage.pragma.user_version(cur_version + 1);
   }
```

Comment:
We can maybe do this without setting all the intermediate versions?

```c++
// ensure update path is possible
auto cur_version = get_version(ctt, storage);
ceph_assert(cur_version <= SFS_METADATA_VERSION);
ceph_assert(cur_version >= SFS_METADATA_MIN_VERSION);
// update
storage.pragma.user_version(SFS_METADATA_VERSION);
```

Reply:
I just really like the idea of updating the db version as the upgrade goes on, in case something fails, so we know at what version we are at any given moment.